vislib gauge chart

ppisljar opened the pull request:
Summary: The gauge and goals charts are two new visualization types that allow people to track a metric and display the in context of a set of reference values. Gauges are useful to indicate how a metric compares to a range of threshold values, for example, to show whether a server load is within a normal range or instead has reach critical capacity. Goal visualizations are similar, but are primarily used to indicate how far a metric is removed from a certain target value.

![image](https://user-images.githubusercontent.com/1833023/27455474-64757d20-576b-11e7-9fde-075ffa4543b4.png)

tbragin commented:
@ppisljar Looking good in terms of functionality! I like what you did with appearance as value passes different thresholds. 

We may be at a point of just tweaking the defaults... My 2c on that: 
* I think Greed-Yellow-Red scheme should be the default as opposed to Yellow-Orange-Red (see screenshots). 
* Btw, i find it a bit annoying that color scheme resets every time I adjust the values in existing ranges, w/o even changing the number of ranges. Is that something we could change?
* We already offer configurability of widths of line and back. Should the gauge color band width also be configurable?
* I find the Percentage Mode a bit unintuitive. What is it a percentage of? Should we print current value and the max from which we calculate this percentage? Or perhaps show it in a mouseover? I'm on the fence whether Percentage Mode makes sense to be on by default. Thoughts?
* I'm sure you already know, but support for field formatters would be great. 
* We need some sort of labelling to distinguish between gauges split by term. 

<img width="1437" alt="screen shot 2017-02-20 at 11 55 08 am" src="https://cloud.githubusercontent.com/assets/933397/23139967/f03b6a1c-f763-11e6-8993-590a923e1b93.png">
<img width="1440" alt="screen shot 2017-02-20 at 11 55 26 am" src="https://cloud.githubusercontent.com/assets/933397/23139968/f03da2c8-f763-11e6-93ff-30de71575871.png">
<img width="1438" alt="screen shot 2017-02-20 at 11 55 59 am" src="https://cloud.githubusercontent.com/assets/933397/23139969/f03f8160-f763-11e6-927f-0853738b1592.png">

<img width="1440" alt="screen shot 2017-02-20 at 12 02 32 pm" src="https://cloud.githubusercontent.com/assets/933397/23140121/addb9088-f764-11e6-9abe-d0802b449712.png">
<img width="1439" alt="screen shot 2017-02-20 at 12 03 15 pm" src="https://cloud.githubusercontent.com/assets/933397/23140122/ade01586-f764-11e6-8a56-0c206e9f96bb.png">

I think it's a good replacement for Kibana 3 goal panel. However, going back to one of the previous points, ability to customize labels to show current number and max number would be good. 

<img width="1438" alt="screen shot 2017-02-20 at 12 11 26 pm" src="https://cloud.githubusercontent.com/assets/933397/23140353/ec827698-f765-11e6-8037-80416ecd2b94.png">
<img width="335" alt="screen shot 2017-02-20 at 12 11 11 pm" src="https://cloud.githubusercontent.com/assets/933397/23140351/ec717b4a-f765-11e6-8323-7b893f5c407d.png">

ppisljar commented:
- i added Sub Text under style, where you can enter things like `of 30` ... but you can put any text there.
- field formatters are not working
- chart title appears (specially useful with split chart)
- Split Group (next to split chart) ... we probably want to keep just one. play with them and let me know what you prefer
- configurable scale width

TODO:
- [ ]  issues with long labels overflowing the chart
- [ ]  percentage mode should show percentages in legend and tooltips as well.
- [ ] UI review
- [ ]  code style
- [ ]  tests
- [ ]  docs

about resetting of color scheme .... i don't think i can do anything there (unless with doing some ugly hack). the problem is in how we handle colors (colors are per label and are used globally). something to change in the future imo.

about percentage mode: i followed what i saw in other tools (splunk last time, tableau goes with the same).
- the expectation is somehow that your data will be in the range from 0 to 100 (thats the default range setup) ... so in this mode you know exactly what the percentage is then ....
- if you setup custom ranges ... then i guess you also know what it means ....
- i find non percentage mode  less ideal for gauges (gauge is filling toward some random number ? i prefer it going toward 100%), thats why i think percentage should be the default (and because others are doing it) .... however im ok with turning it off if there is strong oposition against it.

ppisljar commented:
regarding overflowing labels: with other charts we would usually hide the labels when they overflow. should we do the same here ?

alexfrancoeur commented:
I agree with @ppisljar that percentage mode should be the default. When I think of a gauge, I think of it terms of 0 - 100%. I believe @tbragin's suggestion might be a good one here. Rather than defaulting to 100% we could default to the max number that is available for that value at the time of creation. This would make it a lot easier to see "percentage of total" when splitting into other buckets. If you're dealing with large numbers, it is a bit confusing to see 1M % when defaulting to 100.

Additionally, it might be nice to have an option to create a gauge dynamically using the maximum values and a set number of ranges. This would essentially be a "zero configuration" gauge. This could be more of a future task.

Somewhat related to that, if you do exceed 100% is there a way we can visualize that in the chart? The number might be obvious enough but at first glance, the gauge looks like it's at the max - not over it. Something like a thick bar at the end of the chart might be helpful here - this is just a thought and might not be preferable to some users.
<img width="1386" alt="screen shot 2017-02-21 at 9 10 18 am" src="https://cloud.githubusercontent.com/assets/25181823/23168703/114aab24-f817-11e6-8fab-fb0ae862a74d.png">

I imagine users will want to see the text and percentage within the gauge. What if we added options to configure the title size, number size and possibly padding from the chart? You can see the place names here are very close to the gauges.
<img width="1087" alt="screen shot 2017-02-21 at 9 29 06 am" src="https://cloud.githubusercontent.com/assets/25181823/23169040/48f83a36-f818-11e6-98ed-0f091304a66a.png">

I came across two issues that I think might require "inline errors" similar to derivatives. Or at least some type of warning to a user. Let me know what you think.

Adding a range with a lower "to" value than "from". Some context around this before or after hitting the play button would be useful. Otherwise your gauge looks like this.
<img width="1228" alt="screen shot 2017-02-21 at 9 25 08 am" src="https://cloud.githubusercontent.com/assets/25181823/23168916/d872bc46-f817-11e6-94d5-18eb994f6abc.png">

Should we block a user from creating a gauge if they do not define the range? Otherwise you end up with "Infinity %"
<img width="1433" alt="screen shot 2017-02-21 at 9 11 09 am" src="https://cloud.githubusercontent.com/assets/25181823/23168970/08eef0d8-f818-11e6-9cad-f256e14fe51e.png">

Hope this helps. Happy to clarify over zoom if necessary.

ppisljar commented:
thanks a lot @alexfrancoeur, my comments below:
- i agree on all of your suggestions about the 'zero config chart', however we should definitely split it into separate PR as that on itself might be quite some code and discussions around implementation and usability.
- exceeding 100% .... don't have anything good and simple in mind. not sure what you mean with thick bar, could you post some sample image ?
- regarding font configuration .... i would like to keep that out at this point. I am still thinking of a good way to bring that in across whole visualize, as it would be useful pretty much everywhere.
- however i agree that currently text is to close to the bar ... need to fix that. percentage will always be visible ... if you use non percentage mode number might exceed the boundaries (depends on the size of your chart ..... and your screen) and i think in this case we should hide in and indicate that we did in similar manner we do with heatmap.
- we should define telly enforce correct range definition. i'll work on that.

ppisljar commented:
![gague11](https://cloud.githubusercontent.com/assets/13629809/23210998/31d47ac0-f900-11e6-94d2-3a3fe6543e80.png)

ppisljar commented:
![gauge22](https://cloud.githubusercontent.com/assets/13629809/23211119/bd56d386-f900-11e6-9569-c4658309babd.png)

alexfrancoeur commented:
@ppisljar 

Zero config gauge
 * This was just something that came to mind as I was messing with all the configurations. Would you like me to create an issue?
Exceeding 100%
* My original thought was something along these lines. As I look at it, it doesn't seem as obvious either.
![over 100](https://cloud.githubusercontent.com/assets/25181823/23216351/a4a1a54a-f8e3-11e6-8b5d-42245c8fdffc.png)

I did a bit of digging and there are some other interesting approaches out there. An interesting one I found was on [JS Fiddle.](http://jsfiddle.net/fusioncharts/Lxz2T/) They basically extend the range to be the maximum value if it exceeds. 

<img width="412" alt="screen shot 2017-02-22 at 9 47 57 am" src="https://cloud.githubusercontent.com/assets/25181823/23216483/19f53212-f8e4-11e6-87f3-c47cc5c82071.png"> <img width="404" alt="screen shot 2017-02-22 at 9 48 27 am" src="https://cloud.githubusercontent.com/assets/25181823/23216491/205bb70c-f8e4-11e6-9ee7-63b8ab55f353.png">

Salesforce seems to take a similar approach
<img width="278" alt="screen shot 2017-02-22 at 9 59 35 am" src="https://cloud.githubusercontent.com/assets/25181823/23216943/ab0432b6-f8e5-11e6-9500-acdb2b9a2035.png">

Splunk visualizes this with an animation. Though I could see how this could be annoying on a dashboard
![feb-22-2017 09-54-18](https://cloud.githubusercontent.com/assets/25181823/23216716/f8e2ff4a-f8e4-11e6-9149-f947ad058b52.gif)

General feedback:
 * I love what you did with the ranges! Taking the initial difference and incrementing on it each time you hit "add range" makes life a lot easier - great idea 👍 Splunk seems to do this as well.
* Error's don't necessarily need to be more intuitive for ranges. Eventually I'd like to see more descriptive text for all errors, but that's a far larger task. This should work
<img width="355" alt="screen shot 2017-02-22 at 10 02 20 am" src="https://cloud.githubusercontent.com/assets/25181823/23217096/21bc7968-f8e6-11e6-9258-4fc4af01593c.png">
<img width="350" alt="screen shot 2017-02-22 at 10 03 49 am" src="https://cloud.githubusercontent.com/assets/25181823/23217125/3ccc68a8-f8e6-11e6-9d08-c59904644432.png">
 * The text spacing also looks much better

alexfrancoeur commented:
@ppisljar this seems somewhat related to a more dynamic approach to ranges
https://github.com/elastic/kibana/issues/10459

ppisljar commented:
@alexfrancoeur seems this issue covers our zero config chart partly as well .... do you want to add some more information to this ticket or open a new one ? i think the auto extending range when we go over 100% also falls under this category, so i would keep this for another PR as well.

im gonna schedule UI review for tomorrow so some more ppl can pitch in then we can decide on monday next week where to take this.

alexfrancoeur commented:
@ppisljar I'll update https://github.com/elastic/kibana/issues/10459 with additional info on zero config + auto extending the range when > 100%. Looking forward to the UI review.

tbragin commented:
@alexfrancoeur @ppisljar The reason I am dubious about making percentage mode the default is because it's practically impossible to calculate a ratio over time series right now in Visualize. Some people may calculate these ratios in the fly outside of Elasticsearch and index them in directly, but it's the minority. Until we have something like [Filter Ratio function](https://github.com/elastic/kibana/pull/9725#issuecomment-280956527) in Timeseries visual builder implemented in Visualize, I'd caution against percentage mode being default for gauge.

ppisljar commented:
@tbragin i don't follow ... this is generally a metric, not so much a time series ... i was playing with some metric beat data and there some data is already in percentages. My look on this is that this way for some users we might already show the right chart ... most of them will need to configure it. If we just show the maximum value as 100% ... everyone will need to reconfigure it.

Lets discuss this in the UI review ...

tbragin commented:
@ppisljar Tried out this PR with the metric vis. Looks pretty good! A couple of oddities I noticed (I'm sure this is still WIP, so perhaps this is already on your radar), as well as some general suggestion.

* Splitting by terms sometimes overlaps values
<img width="1428" alt="screen shot 2017-05-03 at 8 14 12 pm" src="https://cloud.githubusercontent.com/assets/933397/25689555/0e156098-303e-11e7-8f0a-ec7d826973da.png">

* When coloring is not used, what does the legend really represent? Can we hide it unless someone has turned on using colors? Same with range options.
<img width="1420" alt="screen shot 2017-05-03 at 8 24 10 pm" src="https://cloud.githubusercontent.com/assets/933397/25689593/7f4642f0-303e-11e7-82a7-d896dbf65d3e.png">

* Splitting by terms horizontal vs vertical shows different number of terms (horizontal is one short)
<img width="1427" alt="screen shot 2017-05-03 at 8 14 32 pm" src="https://cloud.githubusercontent.com/assets/933397/25689561/29266c24-303e-11e7-8dd4-eb2578e3be47.png">
<img width="1427" alt="screen shot 2017-05-03 at 8 23 30 pm" src="https://cloud.githubusercontent.com/assets/933397/25689589/6999fd34-303e-11e7-9232-44ca45861af2.png">

* Ideal starting coloring for operations would assume three ranges and the following coloring scheme. 
<img width="1424" alt="screen shot 2017-05-03 at 8 19 12 pm" src="https://cloud.githubusercontent.com/assets/933397/25689599/927b8dbc-303e-11e7-947b-b1b9cfde5771.png">

* Is the "Back/Text Color" config necessary?

* Is the "Sub text" config necessary?

* Font size adjustor doesn't seem to do anything.

tbragin commented:
For circle vis type (aka circular gauge):

* Longer labels (e.g. URLs) will always be hidden due to size constraints -- there is just not that much space inside a circle :)  Should we provide an option to put those labels outside the circles (e.g. on the bottom)?
<img width="1426" alt="screen shot 2017-05-03 at 8 29 22 pm" src="https://cloud.githubusercontent.com/assets/933397/25689761/d8e27f94-303f-11e7-9fb3-7775da2a182d.png">

ppisljar commented:
@tbragin thanks! regarding the circle vis type label ... should that be an option or should we try to automatically handle the long labels better ? if its an option, what should be the default ? maybe 'outside of the circle' since its gonna work most of the time ?

rasroh commented:
Built from  this PR https://github.com/elastic/kibana/pull/10336 

When I spin up Kibana to see the gauge viz, on the Visualization landing page I see `Gauge` and `goal` which are the same visualizations. Talked to Thomas about it and  `goal` should be removed.  
<img width="1389" alt="screen shot 2017-05-05 at 11 32 37 am" src="https://cloud.githubusercontent.com/assets/11466284/25759337/d74d1232-3186-11e7-83da-247e61856990.png">

rasroh commented:
All the labels are not shown due to container size. Can the size be made bigger ? Else seeing the error message on the screen as shown.  

<img width="1416" alt="screen shot 2017-05-05 at 11 59 55 am" src="https://cloud.githubusercontent.com/assets/11466284/25760183/6e7b9446-318a-11e7-9ea1-e5059cd319cf.png">

rasroh commented:
is it a good idea to add a tool tip to **CIDR Masks.** It would help them understand what that means. 


<img width="1420" alt="screen shot 2017-05-05 at 12 21 08 pm" src="https://cloud.githubusercontent.com/assets/11466284/25760810/5cec5834-318d-11e7-88e5-a5035b81275c.png">

alt74 commented:
new gauge icon in SVG format @thomasneirynck @ppisljar cc @snide @cjcenizal 
[icon-gauge.svg.zip](https://github.com/elastic/kibana/files/980272/icon-gauge.svg.zip)

mockup:
<img width="173" alt="gauge-lg" src="https://cloud.githubusercontent.com/assets/1134701/25763501/73b0d0d0-31b1-11e7-9d1b-8ea19608eec1.png">

thomasneirynck commented:
jenkins, test this

thomasneirynck commented on src/core_plugins/kbn_vislib_vis_types/public/controls/gauge_options.js:
```diff
@@ -0,0 +1,110 @@
+            $scope.vis.params.gauge.minAngle = 0;
+            $scope.vis.params.gauge.maxAngle = 2 * Math.PI;
+            break;
+          case 'Metric':
```
there's a small bug when switching to metric:

```
Error: Accessing invalid config property: fontSize
    at VisConfig.get (https://localhost:5601/atr/bundles/kibana.bundle.js?v=8467:112087:18)
    at SimpleGauge.drawGauge (https://localhost:5601/atr/bundles/kibana.bundle.js?v=8467:322559:59)
    at https://localhost:5601/atr/bundles/kibana.bundle.js?v=8467:322042:40
    at Array.forEach (native)
    at HTMLDivElement.<anonymous> (https://localhost:5601/atr/bundles/kibana.bundle.js?v=8467:322037:26)
    at https://localhost:5601/atr/bundles/kibana.bundle.js?v=8467:96916:17
    at d3_selection_each (https://localhost:5601/atr/bundles/kibana.bundle.js?v=8467:96922:31)
    at Array.d3_selectionPrototype.each (https://localhost:5601/atr/bundles/kibana.bundle.js?v=8467:96915:13)
    at Array.<anonymous> (https://localhost:5601/atr/bundles/kibana.bundle.js?v=8467:322030:22)
    at Array.d3_selectionPrototype.call (https://localhost:5601/atr/bundles/kibana.bundle.js?v=8467:96929:15)
```
To reproduce:
1. choose a gauge-viz
2. switch to metric gauge-type, press play
3. -> screen goes blank

thomasneirynck commented on src/core_plugins/kbn_vislib_vis_types/public/gauge.js:
```diff
@@ -0,0 +1,91 @@
+        name: 'metric',
+        title: 'Metric',
+        min: 1,
+        aggFilter: '!std_dev',
```
in the meantime, we should add geo_centroid
```['!std_dev', '!geo_centroid']```

thomasneirynck commented on src/core_plugins/kbn_vislib_vis_types/public/goal.js:
```diff
@@ -0,0 +1,89 @@
+      }
+    ])
+  });
+}
```
we can delete this file I think (including the reference in kbn_vis_types)

thomasneirynck commented on src/core_plugins/kbn_vislib_vis_types/public/gauge.js:
```diff
@@ -0,0 +1,91 @@
+  return new VislibVisType({
+    name: 'gauge',
+    title: 'Gauge',
+    icon: 'fa-tachometer',
```
@alt74 uploaded a new icon

thomasneirynck commented on src/core_plugins/kbn_vislib_vis_types/public/gauge.js:
```diff
@@ -0,0 +1,91 @@
+    title: 'Gauge',
+    icon: 'fa-tachometer',
+    description: 'A gauge chart shows much more than one value. It gives the minimum, the maximum, ' +
+    'the current value and how far from the maximum you are ',
```
consider rephrasing:

> Gauges to indicate the status of a metric. Use it to show how a metric's value relates to reference threshold values.

thomasneirynck commented on src/ui/public/vislib/visualizations/gauges/gauge_types.js:
```diff
@@ -0,0 +1,10 @@
+
+export function GaugeTypesProvider(Private) {
+
+  return {
```
consider renaming both of these. e.g. `metric` and `dial` ...

thomasneirynck commented on src/ui/public/vislib/visualizations/gauges/meter.js:
```diff
@@ -0,0 +1,362 @@
+      const minAngle = this.gaugeConfig.minAngle;
+      const applyMask = this.gaugeConfig.style.mask;
+      const applyMaskToBack = this.gaugeConfig.style.bgMask;
+      const angleFactor = this.gaugeConfig.gaugeType === 'Meter' ? 0.75 : 1;
```
use the enumeration from `visualizations/gauges/gauge_type.js`

ppisljar replied:
this is actually not a gauge type just a setting for the meter gauge type

thomasneirynck commented on src/ui/public/vislib/visualizations/gauges/simple.js:
```diff
@@ -0,0 +1,208 @@
+    getColorBucket(val) {
+      const bucket = this.getBucket(val);
+      const labels = this.gaugeChart.handler.visConfig.get('legend.labels');
+      return this.colorFunc(labels[bucket]);
```
The color-mapping seems to miss the last color of the defined range. 

E.g., the green-red color range should end with a red color, instead of with an orange.

![image](https://cloud.githubusercontent.com/assets/1833023/25806721/26c13c9c-33d2-11e7-8009-a99df5311e9b.png)

We have the same issue when inverting. The range should begin with red and end with a dark-green color.

![image](https://cloud.githubusercontent.com/assets/1833023/25806760/4a29684e-33d2-11e7-8c17-d5e0dec7c8dc.png)

thomasneirynck commented on src/ui/public/vislib/visualizations/gauge_chart.js:
```diff
@@ -0,0 +1,77 @@
+
+  const Chart = Private(VislibVisualizationsChartProvider);
+  const gaugeTypes = Private(GaugeTypesProvider);
+  const defaults = {
```
could we remove this config? we already have additional default configs for the individual chart types.

thomasneirynck commented on src/ui/public/vislib/visualizations/gauges/meter.js:
```diff
@@ -0,0 +1,362 @@
+      const colorsRange = this.gaugeConfig.colorsRange;
+      const labels = this.getLabels();
+      const colors = {};
+      for (const i in labels) {
```
use array-loop. 
`for (const i = 0; i < labels.length; i += 1)`

can we remove the `if(labels[i])`? I didn't really see that path triggered.

thomasneirynck commented on src/ui/public/vislib/visualizations/gauges/meter.js:
```diff
@@ -0,0 +1,362 @@
+      return this.colorFunc(labels[bucket]);
+    }
+
+    drawScale(svg, data, radius, angle) {
```
unused param `data`

thomasneirynck commented on src/ui/public/vislib/visualizations/gauges/meter.js:
```diff
@@ -0,0 +1,362 @@
+      return scale;
+    }
+
+    drawBarMask(radius, angle, maskBars, maskPadding, name) {
```
unused param `maskPadding`

ppisljar replied:
removing masks all together

thomasneirynck commented:
New icon for the goal visualization from @alt74 

<img width="334" alt="goal-icon" src="https://cloud.githubusercontent.com/assets/1833023/25900508/abd74968-3561-11e7-98e6-65f71f72ce8a.png">

ppisljar commented:
@tanya:
- i think the back/text color config is necessary (you might define colors that won't work with out default)
- subtext is also useful, allows you to better describe your metrics 
- regarding the overlapping text: now i hide the overflow, but thats not ideal solution ... should we:
 -- display same message we do with circle that some labels were hidden ? however here its the value, not just the labels
 -- automatically reduce the font size ? (users might be confused on whats happening)
 -- with circles .. moving under only helps with vertical split ... should we only show the message about hidden labels in editor ? whats your taught @thomasneirynck  (you had similar issues with tagcloud)
other stuff should be fixed

@thomasneirynck i think i updated everything you mentioned

@rasroh regarding the CIDR masks, definetelly, but that applies everywhere not just to gauge, so could you pls create an issue for it?

ppisljar commented:
some unrelated selenium tests to be failing:
fail: "management scripted fields creating and using Painless boolean scripted fields should see scripted field value in Discover" and similar and some tilemap and table vis tests as well ...

thomasneirynck commented:
jenkins, test this

thomasneirynck commented on src/core_plugins/kbn_vislib_vis_types/public/controls/gauge_options.html:
```diff
@@ -0,0 +1,282 @@
+  </div>
+
+  <div class="kuiSideBarFormRow">
+    <label class="kuiSideBarFormRow__label" for="verticalSplit">
```
consider adding little`kbn-info` tooltip with short explanation. It isn't straightforward.

thomasneirynck commented on src/core_plugins/kbn_vislib_vis_types/public/controls/gauge_options.html:
```diff
@@ -0,0 +1,282 @@
+
+  <div class="kuiSideBarFormRow" ng-hide="vis.params.gauge.type === 'simple'">
+    <label class="kuiSideBarFormRow__label" for="extendRange">
+      Auto Extend Range
```
consider adding little`kbn-info` tooltip with short explanation. It isn't straightforward.

thomasneirynck commented on src/core_plugins/kbn_vislib_vis_types/public/controls/gauge_options.html:
```diff
@@ -0,0 +1,282 @@
+
+          <div class="kuiSideBarFormRow">
+            <label class="kuiSideBarFormRow__label" for="backColor">
+              Back Color
```
rename to 'Background color'

ppisljar commented:
@thomasneirynck if we were to hide color options from goal chart then you have no way of going from goal to gauge (and vice versa) ... but i agree that UI is confusing and could use some reorganization.

for now i hide color options if you have just one range defined (as well as some other settings), but i don't like this much either.

kobelb commented on src/ui/public/vislib/visualizations/gauge_chart.js:
```diff
@@ -0,0 +1,71 @@
+  const Chart = Private(VislibVisualizationsChartProvider);
+  const gaugeTypes = Private(GaugeTypesProvider);
+
+  class GaugeChart extends Chart {
```
Occasionally I'm seeing the gauge's render at different places in the screen. For example, Metric visualizations when choosing the Meter Gauge type render here:

<img width="1168" alt="screen shot 2017-05-18 at 11 07 34 am" src="https://cloud.githubusercontent.com/assets/627123/26209243/3ea28ce0-3bba-11e7-84c6-491822aecdf5.png">

And Circle Gauge Types render here:

<img width="1166" alt="screen shot 2017-05-18 at 11 07 42 am" src="https://cloud.githubusercontent.com/assets/627123/26209249/452ab93e-3bba-11e7-8193-7653e20714b1.png">

ppisljar replied:
not sure what you mean ?

kobelb replied:
One of the visualizations is "higher" than the other. It almost looks like one is center aligned, and the other one is bottom aligned.

alexfrancoeur commented:
@ppisljar finally had a chance to sit down and play with this fantastic PR. Feedback below. I apologize if some of it is repetitive.

 * I couldn't find the add range button here. It seems a bit hidden. I was looking for a *+* or some sort of button in general. Maybe design can help here. @cjcenizal thoughts?
<img width="395" alt="screen shot 2017-05-18 at 3 25 48 pm" src="https://cloud.githubusercontent.com/assets/25181823/26219909/b6e64fba-3bde-11e7-85e6-47dc5f6514f8.png">

* There seems to be a discrepancy for percent mode between goal and gauge. I'm charting the same values over the same timeframe which is 0.963. If I switch on percentage mode for both charts I get different values. I would think they should match.

<img width="300" alt="screen shot 2017-05-18 at 3 47 07 pm" src="https://cloud.githubusercontent.com/assets/25181823/26220631/5dda7880-3be1-11e7-862a-68910ea49051.png"> <img width="300" alt="screen shot 2017-05-18 at 3 47 37 pm" src="https://cloud.githubusercontent.com/assets/25181823/26220632/5dedf34c-3be1-11e7-9f50-e51568428e58.png">

<img width="300" alt="screen shot 2017-05-18 at 3 47 25 pm" src="https://cloud.githubusercontent.com/assets/25181823/26220638/63d8be5e-3be1-11e7-83f3-7ff1ccef6b41.png"> <img width="300" alt="screen shot 2017-05-18 at 3 47 42 pm" src="https://cloud.githubusercontent.com/assets/25181823/26220639/63f2de7e-3be1-11e7-9b52-281501a84d6e.png">

* I agree that the info buttons seem to be pretty squished in there, but that's pretty minor

* I find it fairly awkward having the range defined for a metric visualization by default. I thought there was a bug for a moment and then I realized that I had to open up style to select what would be colored. What are your thoughts on this solution? 
  * Introduce an enable / disable option for the metric view
  * If you choose metric through the create visualization page, the option is automatically disabled
  * Otherwise, if you switch between visualization types it's enabled for metric
  * I don't think we can have the enable/disable control for gauge or goal though
My only concern here is that it's still difficult to tie the color directly to the range. This approach does make it a bit less confusing though
<img width="1240" alt="screen shot 2017-05-18 at 4 38 33 pm" src="https://cloud.githubusercontent.com/assets/25181823/26222565/74a14894-3be8-11e7-88bb-805e7c690d76.png">

* We hide derivative in the current metric visualization (5.4). We'll need to hide it with gauge, goal and metric as well. We get an error because the aggregation requires a histogram or date histogram. 

<img width="920" alt="screen shot 2017-05-18 at 3 51 53 pm" src="https://cloud.githubusercontent.com/assets/25181823/26223476/300742f2-3bec-11e7-8aa6-0872c74d45c6.png">

That all being said, I absolutely love having the additional capabilities in metric and the two new visualizations. I can't wait to get them merged!

ppisljar commented on src/core_plugins/kbn_vislib_vis_types/public/controls/gauge_options.html:
```diff
@@ -0,0 +1,294 @@
+
+        <div class="kuiSideBarFormRow">
+          <label class="kuiSideBarFormRow__label" for="invertColors">
+            Reverse Color Schema
```
can't reproduce ... which color schema ?

ppisljar replied:
ah ... only with metric

ppisljar commented on src/core_plugins/kbn_vislib_vis_types/public/controls/gauge_options.html:
```diff
@@ -0,0 +1,294 @@
+    <div ng-if="showStyle" class="kuiSideBarCollapsibleSection">
+      <div class="kuiSideBarFormRow" ng-show="vis.params.gauge.colorsRange.length > 1">
+        <label class="kuiSideBarFormRow__label" for="labelColor">
+          Color Labels
```
a bug in vislib, where you can't initially hide legend :(

ppisljar replied:
this is now a dropdown with 3 values: color labels, color background or color none
hope this makes it clearer ...

ppisljar commented:
@alexfrancoeur:
- the percent mode calculates percentage based on your range ... in your first vis your range is from 0 to 10000 and in second one its from 0 to 100 ... 
- for now i just removed the 3 default ranges with metric and used 1 ... 
- @cjcenizal what should i do with the info icon ? add margins to it ?

cjcenizal commented:
@alexfrancoeur Re the button, I'd just use a Basic Button with the "Small" modifier applied, with the label "+ Add range".

@ppisljar Can you add a non-breaking space between the label and the info button? `&nbsp;`

alexfrancoeur commented:
@ppisljar that's my mistake on the percentage mode - good catch. The experience for ranges in a metric visualization still isn't ideal, but I don't want this to be a blocker for 5.5

It looks like on the metric visualization we are still showing meter instead of arc and it is causing some issues.
![may-23-2017 10-22-11](https://cloud.githubusercontent.com/assets/25181823/26358861/de54c074-3fa1-11e7-8711-52204e0a944f.gif)

@cjcenizal basic button looks good! much more noticeable
<img width="357" alt="screen shot 2017-05-23 at 10 19 09 am" src="https://cloud.githubusercontent.com/assets/25181823/26358687/48a817ba-3fa1-11e7-88e2-1b05e75f7c2a.png">

cjcenizal commented:
@alexfrancoeur Almost a little too noticeable! :) @ppisljar Could you change that to use the `kuiButton--basic` modifier? In general we want to reserve the `primary` modifier for CTAs, e.g. a submit button on a form.

ppisljar commented:
@cjcenizal i think it should match what we have elsewhere when it comes to range definitions (range bucket aggregation, heatmap)

ppisljar commented:
@kobelb i think this is ready for another look

cjcenizal commented:
@ppisljar OK, we can leave it for now. We'll just need to update all of those buttons eventually. Would you mind making an issue for this?
